src/mlpack/methods/mean_shift/mean_shift.hpp

```diff
@@ -0,0 +1,116 @@
+/**
+ * @file mean_shift.hpp
+ * @author Shangtong Zhang
+ *
+ * Mean Shift clustering
+ */
+
+#ifndef __MLPACK_METHODS_MEAN_SHIFT_MEAN_SHIFT_HPP
+#define __MLPACK_METHODS_MEAN_SHIFT_MEAN_SHIFT_HPP
+
+#include <mlpack/core.hpp>
+#include <mlpack/core/kernels/gaussian_kernel.hpp>
+
+namespace mlpack {
+namespace meanshift /** Mean Shift clustering. */ {
+
+/**
+ * This class implements Mean Shift clustering.
+ * For each point in dataset, apply mean shift algorithm until maximum 
+ * iterations or convergence. 
+ * Then remove duplicate centroids.
+ * 
+ * A simple example of how to run Mean Shift clustering is shown below.
+ *
+ * @code
+ * extern arma::mat data; // Dataset we want to run Mean Shift on.
+ * arma::Col<size_t> assignments; // Cluster assignments.
+ * arma::mat centroids; // Cluster centroids.
+ *
+ * MeanShift<arma::mat, kernel::GaussianKernel> meanShift();
+ * meanShift.Cluster(dataset, assignments, centroids);
+ * @endcode
+ *
+ * @tparam KernelType the kernel to use.
```

Comment:
Maybe kernels can be tagged "translation_invariant" as a type trait? Then you can enforce this at compile time